Use msds-memberOfTransitive when possible for AD group lookups

ncstate-mafields opened the pull request:
This PR makes use of the `msds-memberOfTransitive` computed property for AD group lookups when possible, instead of the recursive group lookups using `memberOf`. It still falls back to recursive lookups if the transitive property is not available, or if anything fails while attempting to use the transitive property. Therefore it should be backwards compatible with older domains. 

The [msds-memberOfTransitive](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-adls/dcddfdd8-5c70-45b7-bc64-a714ed2ca621) property was added in Microsoft Server 2019 R2. We can check the [domainFunctionality](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-adts/6dd88965-8feb-4369-ae7e-075985da8071) property to check if the value is greater than or equal to 6 (`DS_BEHAVIOR_WIN2012R2`). 

I've tested this on my own Foreman server, and it works well. Tests should also be passing. 

Fixes #68

lzap commented on lib/ldap_fluff/ad_member_service.rb:
```diff
@@ -8,10 +8,32 @@ def initialize(ldap, config)
+    rescue
+      # If any exceptions are raised, fall back to recursive lookup
+      _groups_from_ldap_data(user_data)
+    end
```
Generally, it's not the best practice to utilize exception for code path. See, if there was a network failure during the initial two search calls, the code will fallback and perform another call silently only after than it errors out. At minimum, you should probably define your own exception class and rescue it explicitly, maybe even avoiding exceptions altogether - they are also quite slow compared to if-else.

@ezr-ondrej thoughts?

lzap commented:
Matt, sorry for the delay. Thanks for the patch.

Would you mind updating the README with the new behavior too?

lzap commented:
Robocop is failing, therefore tests cannot be executed.

lzap commented:
Also would you mind dropping a unit test too?

ncstate-mafields commented:
I removed the exception handling in favor of some basic conditional logic, fixed rubocop, and added/updated tests. 

Also added a note in the README as requested.

ncstate-mafields commented:
@lzap Do these changes look good?

lzap commented on test/lib/ldap_test_helper.rb:
```diff
@@ -85,8 +85,8 @@ def ad_group_dn(name = 'group')
-    [{ :memberof => [ad_group_dn] }]
+  def ad_user_payload(name)
+    [{ :memberof => [ad_group_dn], :distinguishedname => [ad_user_dn(name)] }]
   end
```
Can you please make the `name` argument optional and keep at least one old test that makes the query without `distinguishedname`? I want to make sure the code works for old code too.

lzap reviewed (changes requested):
Two more nitpicks.

Apologies for late reply, will do my best to re-review promptly.

ncstate-mafields commented:
Adding a default value for name allowed the original tests to run unmodified except for the addition of the domain functionality value. Great call on that. 

I think I have successfully memoized the domain functionality value to avoid multiple lookups. 
Ruby is not my primary language, so let me know if there is a better way to do it.

lzap commented on lib/ldap_fluff/ad_member_service.rb:
```diff
@@ -8,10 +8,34 @@ def initialize(ldap, config)
+      @domain_functionality = search.first[:domainfunctionality].first.to_i
+    end
+
+    @domain_functionality
```
There are several better ways of doing it in Ruby, but I value more it works fine, this repo has a lot of code closer to Python :-)

lzap reviewed (approved):
I am not gonna nitpick, it looks safe, it has tests.

I am assuming you would like to see a release soon, I do not have permissions so I will need to look around and ask.

lzap commented:
Package update for nightly: https://github.com/theforeman/foreman-packaging/pull/6829
